[IMP] base, mail: turns attachments into links when mail is too heavy

pyduf opened the pull request:
The email was not sent when it was too heavy for the outgoing server. This
implementation solves the problem by allowing to specify (and even autodetect
when supported) a maximum size for email for each outgoing server. If that
limit is reached for an email, its attachment are automatically turned into
downloadable links inserted at the end of the email so that it can be sent and
the recipient can download the attachment through the provided links.

Note that there could still be an error if the recipient server cannot handle
the email size. But it can be mitigated by setting a low value for the max
email size of the outgoing servers.

Limitation: the attachments are only turned into links if they are owned by a
business model. The goal is to delegate the attachment life cycle to the user;
when the attachment is owned by a record, the user can delete it through the
chatter of that record. And of course if the user delete the attachment, the
related link in the email will lead to a 404.

Technical note: You can specify the maximum email size on each outgoing server
(ir.mail_server) and for the smtp server specified on the command line with
the added parameter max-outgoing-email-size (default 5MB).

Task-2474726

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/96008).

std-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -511,11 +540,13 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
                 # TDE note: could be great to pre-detect missing to/cc and skip sending it
                 # to go directly to failed state update
                 for email in email_list:
+                    body_with_embedded_attachment_links = email.get('body') if not attachments_links \
+                        else tools.append_content_to_html(email.get('body'), attachments_links, plaintext=False)
```
a bit more readable imo
```suggestion
                    if attachments_links:
                        email['body'] =   tools.append_content_to_html(email.get('body'), attachments_links, plaintext=False)
```

pyduf replied:
Yes good idea, thanks

std-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -115,6 +116,14 @@ class IrMailServer(models.Model):
     smtp_debug = fields.Boolean(string='Debugging', help="If enabled, the full output of SMTP sessions will "
                                                          "be written to the server log at DEBUG level "
                                                          "(this is very verbose and may include confidential info!)")
+    max_email_size = fields.Float(string="Max email size (MB)", default=5, store=True,
```
it's stored by default
```suggestion
    max_email_size = fields.Float(string="Max email size (MB)", default=5,
```

pyduf replied:
Indeed, thanks

std-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -480,6 +484,31 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
 
+                # Compute email size and transform attachment in downloadable link in the body if it exceeds the limit
+                estimated_email_size = (
+                        len(bytes(json.dumps(headers), "utf-8")) * BASE_64_SIZE_OVERHEAD_RATIO +
+                        len(bytes(body, "utf-8")) * BASE_64_SIZE_OVERHEAD_RATIO +
+                        sum(attachment['file_size'] for attachment in
+                            attachments.sudo().read(['file_size'])) * BASE_64_SIZE_OVERHEAD_RATIO +
```
why do you need to multiply by a base64 ratio ?

(if really needed I would do this instead: (a + b + c) * ratio)

pyduf replied:
Actually this is the worst case, I think. There are different encoding and as far as I know base64 is the worst of them: 6 bits (64 characters) are encoded in 8 bits which leads to a ratio of 8/6. I have factored it as suggested.

std-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -20,6 +21,10 @@
 _logger = logging.getLogger(__name__)
 
+BASE_64_SIZE_OVERHEAD_RATIO = 8 / 6
+EMAIL_SIZE_SECURITY_MARGIN_BYTES = 50 * 1024
```
shouldn't be this in  percent of the limit ? (e.g. for extreme cases, if you server supports only 30ko, it's strange to use a security margin of 50ko)

pyduf replied:
This is for the estimation of the email size to take into account the format that glue all part together (headers, body, attachment). I have reduced it to 10Ko. I think it is small enough regarding the size commonly authorized: almost always 10Mo. The only thing it will cause is to set the attachment as link in the email while it could have sent it as regular attachment because of an over-estimation of 10Ko.

std-odoo reviewed (commented):
@pyduf Hi :) 

Good work, I added some comments / propositions :)

pyduf commented on addons/mail/__manifest__.py:
```diff
@@ -51,6 +51,7 @@
         'views/res_partner_views.xml',
         'views/mail_blacklist_views.xml',
         'views/mail_menus.xml',
+        'views/mail_mail_templates.xml',
```
Yes indeed, it's done.

pyduf commented on addons/mail/models/mail_mail.py:
```diff
@@ -480,6 +484,31 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
+                        sum(attachment['file_size'] for attachment in
+                            attachments.sudo().read(['file_size'])) * BASE_64_SIZE_OVERHEAD_RATIO +
+                        EMAIL_SIZE_SECURITY_MARGIN_BYTES
+                )
```
Actually, I get the error: "TypeError: string argument without an encoding" if I don't set it. Maybe it has changed ?

pyduf commented on addons/mail/views/mail_mail_templates.xml:
```diff
@@ -0,0 +1,18 @@
+            <t t-foreach="attachments" t-as="attachment">
+                <div style="border: 1px solid #C9CCD2; border-radius: 3px; background-color: #C9CCD2; padding:5px; margin:5px">
+                    <a t-attf-href="{{attachment.get_base_url()}}/web/content/{{attachment.id}}?download=1&amp;access_token={{attachment.access_token}}" style="color: black; text-decoration: none;">
+                        <t t-esc="attachment.name"/>
```
Changed

pyduf commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -30,6 +30,7 @@
 _test_logger = logging.getLogger('odoo.tests')
 
 SMTP_TIMEOUT = 60
+BYTES_PER_MB = 1024 * 1024
```
I have removed and inlined the computation (1024 ** 2). I think both way have advantages and disadvantages.

pyduf commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -202,7 +217,10 @@ def test_smtp_connection(self):
+        if is_set_max_email_size:
+            message = _('Max email size has been set')
+        else:
+            message = _("Connection Test Successful!")
```
Yes, I have inlined directly in the dictionary value so the variable message is not needed anymore.

pyduf commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -151,7 +160,7 @@ def _get_test_email_addresses(self):
         return email_from, 'noreply@odoo.com'
 
-    def test_smtp_connection(self):
+    def test_smtp_connection(self, is_set_max_email_size=False):
```
Yes indeed, it's done.

pyduf commented:
> @pyduf Hi :)
> 
> Good work, I added some comments / propositions :)

Thanks for the review. I've just updated the code following your suggestions.

std-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -423,18 +427,23 @@ def send(self, auto_commit=False, raise_exception=False):
                     raise_exception=raise_exception,
-                    smtp_session=smtp_session)
+                    smtp_session=smtp_session,
+                    max_email_size_in_bytes=max_email_size * 1024 ** 2
```
:disguised_face: 
```suggestion
                    max_email_size_in_bytes=max_email_size * 1024 ** 2,
```

pyduf replied:
"," added

std-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -481,6 +484,31 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
 
+                # Compute email size and transform attachment in downloadable link in the body if it exceeds the limit
+                estimated_email_size = (
+                        (len(bytes(json.dumps(headers), "utf-8")) +
```
I don't really get why you need to convert to byte to compute the length, it should be (almost) always the same isn't it ?
```suggestion
                        (len(json.dumps(headers)) +
```

pyduf replied:
Good catch, json.dumps encodes everything by default in ascii making a non ascii text very long; each non ascii characters takes 5 times more place! (instead of 2 times most of the time, see example below).

So the idea is to set ensure_ascii=False to have a better estimation. After the dump we still need to encode it in UTF-8 in order to count the bytes and not the characters. Indeed the goal is to evaluate the message size in byte and not in characters. For example, "é" will takes 2 bytes.

Here we assume that the worst/common encoding is to use utf-8 + base64.

Note that bytes('é', 'utf-8') is the same as 'é'.encode(). But indeed ".encode()" is easier so I have replaced it.

Detailed example: 

{'k': 'é'} --> json.dumps(text) --> {"k": "\u00e9"} --> len=15
        --> print(len(json.dumps(text).encode(errors='ignore'))) = 15
        
{'k': 'é'} --> json.dumps(text, ensure_ascii=False) --> {"k": "é"} --> len=10
        --> len(json.dumps(text, ensure_ascii=False).encode(errors='ignore')) = 11

std-odoo commented on addons/mail/tests/test_mail_composer.py:
```diff
@@ -78,3 +81,65 @@ def test_mail_mass_mode_compose_with_mso(self):
+        """ Test that when mail size exceed the max email size limit, attachment are turned into download links added
+        at the end of the email content. """
+        attachment_size_in_bytes = 1024 * 128
+        datas = base64.b64encode(bytes(''.join('.' for _ in range(0, attachment_size_in_bytes)), 'utf-8'))
```
I really think it should work like this (without the encoding)
```suggestion
        datas = base64.b64encode(''.join('.' for _ in range(0, attachment_size_in_bytes).encode())
```

(same for other occurences)

pyduf replied:
I have replaced bytes('xx', 'utf-8') by 'xx'.encode() as it is equivalent and easier.

std-odoo commented on addons/test_mail_full/tests/test_mail_performance.py:
```diff
@@ -8,7 +8,7 @@
 from odoo.tools import mute_logger
 
 
-@tagged('mail_performance', 'post_install', '-at_install')
```
why this change ?

pyduf replied:
Good catch, sorry, it must have occurred during a rebase.

std-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -202,17 +226,22 @@ def test_smtp_connection(self):
             'params': {
-                'message': message,
+                'message': _('Max email size has been set') if is_set_max_email_size
+                else _("Connection Test Successful!"),
```
:disguised_face: 
```suggestion
                else _('Connection Test Successful!'),
```

pyduf replied:
Updated

std-odoo reviewed (commented):
@pyduf Hi :) 

Some more comments / questions, look almost good :)

pyduf commented:
> @pyduf Hi :)
> 
> Some more comments / questions, look almost good :)

Thanks for the useful remarks.

pyduf commented on odoo/tools/config.py:
```diff
@@ -224,6 +224,9 @@ def __init__(self, fname=None):
         group = optparse.OptionGroup(parser, "SMTP Configuration")
         group.add_option('--email-from', dest='email_from', my_default=False,
                          help='specify the SMTP email address for sending email')
+        group.add_option('--max-outgoing-email-size', dest='max_outgoing_email_size', my_default=5,
+                         help='specify the maximum size (MB) of outgoing mail; when reached, the attachments are sent as links added in the mail body',
+                         type="float")
```
No I haven’t. The reason this parameter has been added is that we can define an outgoing server with the CLI parameters which is not configurable from the interface. To support also the new feature for outgoing server defined with the CLI parameters, we have decided to add that parameter. An alternative would be to not support that new feature for outgoing server defined with the CLI i.e. always send the attachment (not the links). What do you think @antonylesuisse ?

Xavier-Do replied:
Why not using an icp?

tde-banana-odoo replied:
For me both are good, ICP or config. It makes sense to continue the server config by parameters to have a complete setup, but if @odony is strongly against it, I suggest to fallaback on a config parameter as indicated by @Xavier-Do  .

pyduf commented on addons/mail/models/mail_mail.py:
```diff
@@ -508,6 +536,9 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
                 # to go directly to failed state update
                 for email in email_list:
+                    if attachments_links:
+                        email['body'] = tools.append_content_to_html(email.get('body'), attachments_links,
+                                                                     plaintext=False)
```
Good point, we have indeed discussed and tested the visibility of the download links. Adding the link at the top of the mail has been discussed also but we decided to add them at the bottom. I will ask Luc (our PO) for the exact reason and come back to you.

pyduf commented on addons/mail/models/mail_mail.py:
```diff
@@ -477,6 +480,31 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
+                )
+
+                attachments_links = None
+                if attachments and estimated_email_size > max_email_size_in_bytes:
+                    # Remove attachments and prepare downloadable link to be added in the body
+                    for attachment in attachments:
+                        attachment.generate_access_token()
+                    attachments_links = self.env['ir.qweb']._render('mail.mail_attachment_links',
+                                                                    {'attachments': attachments})
+                    attachments = self.env['ir.attachment']
```
Also a good point, thanks. Mails sent (mail_mail) from the chatter with attachment are marked auto_delete  and then to_delete. They are then processed by the cron “Mail: Email Queue Manager” which delete the mail_mail but left the attachment untouched as it is linked to mail_message. So it seems safe but the comment of the method _postprocess_sent_message seems to says the contrary (“including deleting it completely along with its attachment if the ``auto_delete`` flag of the mail was set.”). What do you think @tde-banana-odoo?

tde-banana-odoo replied:
Mmmh, depends probably of the use case
  * comment -> we delete the mail but keep the mail_message and attachments -> should work
  * mass mailing -> we delete the mail and the underlying mail_message -> annoying

Note that we delete attachments linked to the "mail_message", not the attachments linked to the record itself (see the override of unlink in MailMessage). Classic chatter usage links attachments to the record itself so it should not be an issue in that case (probably to check a bit further to be sure).

When doing a pure mass mailing, indeed we might have an issue, could be great to check the composer in mass mailing mode and mailing application, to see what happens with attachments (model / res_id, as I don't remember all details). Then we can check how to avoid the issue.

pyduf commented on addons/mail/models/mail_mail.py:
```diff
@@ -477,6 +480,31 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
+                # load attachment binary data with a separate read(), as prefetching all
+                # `datas` (binary field) could bloat the browse cache, triggerring
+                # soft/hard mem limits with temporary data.
+                attachments = [(a['name'], base64.b64decode(a['datas']), a['mimetype'])
```
Yes indeed, that’s useless work, thanks. I have changed and tested it.

pyduf commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -121,6 +121,14 @@ class IrMailServer(models.Model):
+                                       "A good value is 5MB. "
+                                       "If a mail size exceeds that limit, its attachments will be turned into "
+                                       "downloadable links appended at the end of the email."
+                                  )
```
Thanks for the improved help, it is updated. 
But I haven’t changed the string as in the form we have: 
Max Email Size	5.00	MB
(the unit is just after the field, instead of being in the label)

pyduf commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -209,7 +217,17 @@ def _get_test_email_addresses(self):
+        - if the connection fails
+        - if is_set_max_email_size and the server doesn't support the auto-detection of email max size
+
+        :param bool is_set_max_email_size: whether to set the max email size
```
Yes it is a bit strange.; the goal was to help by telling it is a boolean but it do the contrary. I have renamed it autodetect_max_email_size. I think it is better. Don’t hesitate if it is still unclear.

pyduf commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -233,6 +251,12 @@ def test_smtp_connection(self):
+                    if max_size:
+                        server.max_email_size = float(max_size) / (1024 ** 2)
+                    else:
+                        raise UserError(_("The server doesn't return the email maximum size."))
```
I have updated it, thanks.

pyduf commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -260,17 +284,22 @@ def test_smtp_connection(self):
             'params': {
-                'message': message,
+                'message': _('Max email size has been set') if is_set_max_email_size
+                else _('Connection Test Successful!'),
```
Also updated, thanks

pyduf commented on addons/mail/models/mail_mail.py:
```diff
@@ -395,21 +399,39 @@ def _prepare_outgoing_list(self):
+
             # load attachment binary data with a separate read(), as prefetching all
-            # `datas` (binary field) could bloat the browse cache, triggering
+            # `datas` (binary field) could bloat the browse cache, triggerring
```
Thanks for notifying that mistake, it's now corrected. 
For info, I haven't changed it on purpose but have reintroduced that spelling error during a rebase (that error is present in 16.0 for example).

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -315,16 +318,17 @@ def _prepare_outgoing_body(self):
         return self.env['mail.render.mixin']._replace_local_links(self.body_html)
 
-    def _prepare_outgoing_list(self):
+    def _prepare_outgoing_list(self, max_email_size_in_bytes):
```
I think it is a bit too much low-level to propagate this parameter ... can't we instead propagate ``mail_server`` from the "send" call, as it makes more sense to me ? Having the outgoing server in the "prepare_outgoing_list" method seems legit for me, what do you think ?

Once we have the mail server we can compute back the email size when necessary.

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -395,21 +399,39 @@ def _prepare_outgoing_list(self):
                     attachments = attachments - self.env['ir.attachment'].browse(list(link_ids))
+
+            # Compute email size and transform attachment in downloadable link in the body if it exceeds the limit
+            estimated_email_size = (
```
Would do a small tool function

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -395,21 +399,39 @@ def _prepare_outgoing_list(self):
+                    (len(json.dumps(headers, ensure_ascii=False).encode(errors='ignore')) +
+                     (len(body.encode(errors='ignore')) if body else 0) +
+                     sum(attachment['file_size']
+                         for attachment in attachments.sudo().read(['file_size']))) * BASE_64_SIZE_OVERHEAD_RATIO +
```
Can't we read attachments before, and then compute email size and email_attachments based on it ? That way we don't read it twice (even if I suppose the cache is working ?)

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -395,21 +399,39 @@ def _prepare_outgoing_list(self):
+            if estimated_email_size > max_email_size_in_bytes:
+                # Remove attachments and prepare downloadable link to be added in the body
+                for attachment in attachments:
+                    attachment.generate_access_token()
```
Note that the method is multi, no need to loop on attachments :)

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -495,18 +517,23 @@ def send(self, auto_commit=False, raise_exception=False):
                     batch.write({'state': 'exception', 'failure_reason': exc})
                     batch._postprocess_sent_message(success_pids=[], failure_type="mail_smtp")
             else:
+                max_email_size = self.env['ir.mail_server'].browse(mail_server_id).max_email_size if mail_server_id \
```
I would move this code in "prepare_outgoing_list" if we propagate the mail_server, that way we hide the low level details in that preparation method, do you think this would be ok ?

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -20,6 +20,9 @@
 
 _logger = logging.getLogger(__name__)
 
+BASE_64_SIZE_OVERHEAD_RATIO = 8 / 6
```
If we do a small tool function, you will be able to add a small docstring to explain all those computation :D

tde-banana-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -216,7 +224,17 @@ def _get_test_email_addresses(self):
+    def test_smtp_connection(self, autodetect_max_email_size=False):
+        """Test the connection and if autodetect_max_email_size, set auto-detected max email size.
+
+        The method raises UserError
```
There is a standard way of indicating exceptions in docstrings using ':raises', like
```
:raises:
    UserError: Explanation here.
```

tde-banana-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -267,17 +291,22 @@ def test_smtp_connection(self):
             'params': {
-                'message': message,
+                'message':
+                    _('Max email size has been set') if autodetect_max_email_size else _('Connection Test Successful!'),
```
Maybe we could always say "Connection Succesful.", then add "Email maximum size updated" is the boolean was set ?

tde-banana-odoo commented on addons/mail/tests/test_mail_composer.py:
```diff
@@ -261,3 +264,66 @@ def test_mail_mass_mode_compose_with_mso(self):
         )
+
+    @users('employee')
+    def test_mail_attachment_to_download_link(self):
```
Maybe we could could also make a bit more low-level tests in "test_mail_mail" file ? Notably checking that we take either the server params, either a config parameter if not set / if we have the ICP, ...

We can keep this test as the "higher level one", indicating it works in notification process.

tde-banana-odoo commented on addons/mail/tests/test_mail_composer.py:
```diff
@@ -261,3 +264,66 @@ def test_mail_mass_mode_compose_with_mso(self):
+                                for attachment_name in [attachment1_name, attachment2_name]]
+
+        def send_mail(max_email_size):
+            self.mail_server_global.max_email_size = max_email_size
```
You could force the mail server in the composer (mail_server_id is normally supported), so that you don't have to update both test servers :)

tde-banana-odoo commented on addons/mail/tests/test_mail_composer.py:
```diff
@@ -261,3 +264,66 @@ def test_mail_mass_mode_compose_with_mso(self):
+        match_download_links = [re.compile(fr'.*<a.*href.*/web/content/([^?]*)\?.*access_token.*>.*{attachment_name}.*</a>', re.DOTALL)
+                                for attachment_name in [attachment1_name, attachment2_name]]
+
+        def send_mail(max_email_size):
```
send_mass_mailing maybe ?

tde-banana-odoo commented on addons/mail/tests/test_mail_composer.py:
```diff
@@ -261,3 +264,66 @@ def test_mail_mass_mode_compose_with_mso(self):
+        at the end of the email content. """
+        attachment_size_in_bytes = 1024 * 128
+        datas = base64.b64encode((''.join('.' for _ in range(0, attachment_size_in_bytes))).encode())
+        attachment1_name = 'attachment1'
```
Could be a bit list-comprehensioned, like
```
base_values = {
    'datas': datas,
    'res_id': self.test_record.id,
    'res_name': 'test',
    'res_model': self.test_record._name,
}
attachments = self.env['ir.attachment'].sudo().create([
    dict(base_values, name=name)
    for name in ('attachment1', 'attachment2')
])

tde-banana-odoo commented on addons/test_mail/tests/test_performance.py:
```diff
@@ -832,7 +832,7 @@ def _patched_unlink(records):
         unlinked_mails = set()
 
-        with self.assertQueryCount(admin=43, employee=43), \
+        with self.assertQueryCount(admin=55, employee=55), \
```
Tha's... quite much, is there any issue with the cache like reading the server each time we send any email ? Or additional reads on attachments ?

tde-banana-odoo commented on addons/test_mail/tests/test_performance.py:
```diff
@@ -793,7 +793,7 @@ def test_complex_mail_mail_send(self):
             'recipient_ids': [(4, pid) for pid in self.partners.ids],
         })
-        with self.assertQueryCount(admin=9, employee=9):
+        with self.assertQueryCount(admin=10, employee=10):
```
Same here, could be interesting to see why it grows with the number of email ?

tde-banana-odoo reviewed (commented):
Zboing, review ! Thanks for the PR, glad to see it coming to an end :) 

Cheers !

tde-banana-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -728,6 +733,86 @@ def test_mail_mail_send_server(self):
+                'attachment_ids': attachments.ids,
+                'body_html': '<p>Test</p>',
+                'email_from': 'test@unknown_domain.com',
+                'email_to': f'user{idx_test}_{idx_mail}@example.com',
```
If you really need unique recipients, maybe use test parameters so that we get rid our idx_test and enumerate ? to lessen a bit boilerplate :)

tde-banana-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -728,6 +733,86 @@ def test_mail_mail_send_server(self):
+        datas = base64.b64encode((''.join('.' for _ in range(0, attachment_size_in_bytes))).encode())
+
+        # max_email_size_inv produces the opposite of what is expected, it is assigned to the mail server not used
+        for idx_test, ((max_email_size, max_email_size_inv, turned_into_links, n_attachment), mail_server, n_mail) in \
```
We generally try to use "expected" in variable naming for expected output, to know it is not an input (like 'expected_is_convert', 'expected_attach_count', ...)

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -427,13 +445,25 @@ def _prepare_outgoing_list(self, recipients_follower_status=None):
                     attachments = attachments - self.env['ir.attachment'].browse(list(link_ids))
+
+            # Compute email size and transform attachment in downloadable link in the body if it exceeds the limit
+            estimated_email_size = self._estimated_email_size(headers, body, [a.file_size for a in attachments.sudo()])
```
Sure this won't prefetch anything unwanted ?

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -427,13 +445,25 @@ def _prepare_outgoing_list(self, recipients_follower_status=None):
+                # Remove attachments and prepare downloadable link to be added in the body
+                attachments.generate_access_token()
+                attachments_links = self.env['ir.qweb']._render('mail.mail_attachment_links',
+                                                                {'attachments': attachments})
```
Note that other calls to attachments are done using sudo... wondering why, because normally either the cron is sending emails (which means root), either a standard user (and access should be checked probably). Don't remember why the sudo is used in this method, but seems weird to have part of the code in sudo and part not in sudo.

tde-banana-odoo reviewed (commented):
Zboing, thanks for update ! Added some polishing comments (notably for tests), but globally seems we are seeing the end of this one.

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -319,6 +319,21 @@ def _parse_scheduled_datetime(self, scheduled_datetime):
+        """
+        return (
+                len(json.dumps(headers, ensure_ascii=False).encode(errors='ignore')) +
+                (len(body.encode(errors='ignore')) if body else 0) +
```
Maybe
```suggestion
                len((body or '').encode(errors='ignore')) +
```

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -319,6 +319,21 @@ def _parse_scheduled_datetime(self, scheduled_datetime):
+        :param list attachments_size: list of attachment size in bytes
+        """
+        return (
+                len(json.dumps(headers, ensure_ascii=False).encode(errors='ignore')) +
```
Maybe
```suggestion
                len(json.dumps((headers or {}), ensure_ascii=False).encode(errors='ignore')) +
```

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -425,21 +442,37 @@ def _prepare_outgoing_list(self, recipients_follower_status=None):
+                lambda a: a.res_model and a.res_id and a.res_model != 'mail.message'):
+            estimated_email_size = self._estimated_email_size(headers, body,
+                                                              [a.file_size for a in attachments.sudo()])
+            max_email_size_in_bytes = ((mail_server and mail_server.max_email_size) or
+                                       self.env['ir.mail_server']._get_default_max_email_size()) * 1024 * 1024
```
Maybe we could define "_get_max_email_size" on ir.mail_server doing both and then use it in a one-liner here, like
```suggestion
            max_email_size_in_bytes = mail_server or self.env['ir.mail_server'])._get_max_email_size() * 1024 * 1024

   ... (on mail server model)
   def _get_max_email_size(self):
       if self.max_email_size:
           return self.max_email_size
       return int(self.env['ir.config_parameter'].sudo().get_param('mail.outgoing_email_size', '5'))
```

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -425,21 +442,37 @@ def _prepare_outgoing_list(self, recipients_follower_status=None):
+        # Turn remaining attachments into links if they are too heavy and
+        # theirs ownerships are business models (i.e. something != mail.message,
+        # otherwise they will be deleted along with the mail message leading to a 404)
+        if business_record_attachments := attachments.filtered(
```
```suggestion
        if business_attachments := attachments.filtered(
```

tde-banana-odoo commented on addons/mail/models/res_config_settings.py:
```diff
@@ -70,6 +70,13 @@ class ResConfigSettings(models.TransientModel):
+        "Outgoing Email Size (MB)",
+        help="Default maximum email size (in MB) before the attachments are sent as links in the email body. "
+             "That value can be overriden for each outgoing server.",
+        config_parameter="mail.outgoing_email_size",
```
Maybe
```suggestion
        config_parameter="mail.default_outgoing_email_size",
```

tde-banana-odoo commented on addons/mass_mailing/models/mail_mail.py:
```diff
@@ -62,11 +62,11 @@ def _prepare_outgoing_body(self):
         view links by email-specific links. Also add headers to allow
         unsubscribe from email managers. """
-        email_list = super()._prepare_outgoing_list(recipients_follower_status)
+        email_list = super()._prepare_outgoing_list(mail_server, recipients_follower_status)
```
I prefer to keep named parameters when calling super, like
```suggestion
        email_list = super()._prepare_outgoing_list(mail_server=mail_server, recipients_follower_status=recipients_follower_status)
```

tde-banana-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -129,6 +131,14 @@ class IrMailServer(models.Model):
     smtp_debug = fields.Boolean(string='Debugging', help="If enabled, the full output of SMTP sessions will "
                                                          "be written to the server log at DEBUG level "
                                                          "(this is very verbose and may include confidential info!)")
+    max_email_size = fields.Float(string="Max Email Size", default=lambda self: self._get_default_max_email_size(),
```
Not sure the field should have a default: more a method that returns the default value if not set (as proposed earlier). That way if you know it you can set it, otherwise just leave it blank and let odoo deal with it.

tde-banana-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -273,17 +303,25 @@ def test_smtp_connection(self):
 
-        message = _("Connection Test Successful!")
+        messages = [_('Connection Test Successful!')]
+        if autodetect_max_email_size and self:
```
Maybe we can be defensive
```suggestion
        if autodetect_max_email_size and self.max_email_size:
```

tde-banana-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -824,6 +831,97 @@ def test_mail_mail_values_unicode(self):
+        # email size is used.
+        for (n_attachment, mail_server, default_email_size, max_email_size, business_attachment), expected_is_links in (
+                # 1 attachment which doesn't exceed max size
+                ((1, using_default_mail_server, 0.3, cst_max_size_always_exceed, True), False),
```
Would avoid using a variable, better be explicit it is void, like
```suggestion
                ((1, self.env['ir.mail_server'], 0.3, cst_max_size_always_exceed, True), False),
```

tde-banana-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -824,6 +831,97 @@ def test_mail_mail_values_unicode(self):
+        attachment_size_in_bytes = 1024 * 128
+        datas = base64.b64encode((''.join('.' for _ in range(0, attachment_size_in_bytes))).encode())
+        # Define some constant to ease the understanding of the test
+        using_specific_mail_server = self.mail_server_domain_2
```
```suggestion
        test_mail_server = self.mail_server_domain_2
```

tde-banana-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -824,6 +831,97 @@ def test_mail_mail_values_unicode(self):
+            attachments = self.env['ir.attachment'].sudo().create([{
+                'name': f'attachment{idx_attachment}',
+                'res_name': 'test',
+                'res_model': self.test_record._name if business_attachment else False,
+                'res_id': self.test_record.id if business_attachment else 0,
```
Won't other attachments be linked to the mail.mail or mail.message (e.g. mass mailing case) ? Shouldn't we mimic that behavior ?

tde-banana-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -824,6 +831,97 @@ def test_mail_mail_values_unicode(self):
+            match_download_links = [
+                re.compile(fr'.*<a.*href.*/web/content/([^?]*)\?.*access_token.*>.*{attachment.name}.*</a>', re.DOTALL)
+                for attachment in attachments]
+            for idx_mail in range(2):
```
Maybe
```suggestion
            for mail, outgoing email in zip(mails, self.emails):
```

tde-banana-odoo reviewed (commented):
Thanks for the update and the explanation ! Added some comments but globally, LGTM.

pyduf reviewed (commented):
Hello @tde-banana-odoo , 

Thanks for the review. I have updated the code and answered your comments. 
 
I had also to increase the query count of one test but as explained in the commit, the incriminated part shouldn’t actually add queries.

awa-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -275,17 +303,28 @@ def test_smtp_connection(self):
-        message = _("Connection Test Successful!")
+        messages = [_('Connection Test Successful!')]
+        if autodetect_max_email_size:
+            if self.max_email_size:
```
Looks like this method is meant to be used in multi-mode so this is not correct. 
Not sure what to do with the messages if that's the case... do we really need a message? ^^

awa-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -275,17 +303,28 @@ def test_smtp_connection(self):
+                messages.append(_('Email maximum size updated to %(size)s MB',
+                                  size=', '.join(f"{server.max_email_size:.2f}" for server in self)))
+            else:
+                raise UserError(_("The server doesn't return the maximum email size."))
```
Why not raise this here above? Looks like other server issues are thrown inside the recordset loop.

awa-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -275,17 +303,28 @@ def test_smtp_connection(self):
 
+    def action_retrieve_max_email_size(self):
+        self.ensure_one()
+        # Don't return the client notification because otherwise the data are not updated on the client side
```
I think this is possible to do both, I've done it recently in: https://github.com/odoo/odoo/blob/master/addons/event_crm/models/event_event.py#L75

(If I understand this comment properly).

awa-odoo reviewed (commented):
Hello @pyduf 

I've been pinged by the wise sage that sits at my right, mentioning that this is now a priority task.

So I've had a look and left some comments / questions.
(Did not check tests yet).

Thanks for your work & cheers!

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -336,6 +336,21 @@ def _parse_scheduled_datetime(self, scheduled_datetime):
     # ------------------------------------------------------
 
+    @api.model
+    def _estimated_email_size(self, headers, body, attachments_size):
```
Would keep it as it allows to hide details (and we can find it / tweak it / override it if necessary).

tde-banana-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -442,21 +459,36 @@ def _prepare_outgoing_list(self, recipients_follower_status=None):
-                (a['name'], base64.b64decode(a['datas']), a['mimetype'])
-                for a in attachments.sudo().read(['name', 'datas', 'mimetype']) if a['datas'] is not False
-            ]
+            email_attachments = [(a['name'], a['raw'], a['mimetype'])
+                                 for a in attachments.sudo().read(['name', 'raw', 'mimetype'])
+                                 if a['raw'] is not False]
```
Prefetching issue mainly.

awa-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -336,6 +336,21 @@ def _parse_scheduled_datetime(self, scheduled_datetime):
     # ------------------------------------------------------
 
+    @api.model
+    def _estimated_email_size(self, headers, body, attachments_size):
```
Then let's rename "_estimate_email_size".

awa-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -316,6 +317,21 @@ def _parse_scheduled_datetime(self, scheduled_datetime):
+                len(json.dumps(headers, ensure_ascii=False).encode(errors='ignore')) +
+                (len(body.encode(errors='ignore')) if body else 0) +
+                sum(attachments_size) * 8 / 6 +  # 8 / 6 factor because base 64 are 6 bits coded in 8 bits
+                10 * 1024  # adding 10Ko as security
```
Would add a short version of that explanation in the method docstring, that way we know why we add that amount.

awa-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -367,11 +382,13 @@ def _personalize_outgoing_body(self, body, partner=False, recipients_follower_st
         is a dictionary for specific email values, depending on a partner, or
         generic to the whole recipients given by mail.email_to.
 
+        :param mail_server: mail_server that will be used to send the mails,
```
Could mention the model, like done in other docstrings: `:param partner: <res.partner> recipient`

awa-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -206,6 +207,11 @@ def _active_usages_compute(self):
+    def _get_max_email_size(self):
+        if self.max_email_size:
+            return self.max_email_size
+        return float(self.env['ir.config_parameter'].sudo().get_param('base.default_max_email_size', '5'))
```
Still 5 here, shouldn't we update to 10 as well?

awa-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -275,17 +296,28 @@ def test_smtp_connection(self):
+        messages = [_('Connection Test Successful!')]
+        if autodetect_max_email_size:
+            if all(server.max_email_size for server in self):
+                messages.append(_('Email maximum size updated to %(size)s MB',
+                                  size=', '.join(f"{server.max_email_size:.2f}" for server in self)))
```
As we're making a list of messages anyway, can't we have one message per server? "[Server name] email maximum size was updated to X MB" ?

Otherwise it's impossible to know which server was updated to what.

But honestly it's becoming very complex, as the form view will be updated with the new value now, do we really need to send that notification? We can simplify and directly raise if not supported above, the same way it's done for other issues.

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+        controlled by the user (might even be deleted right after the message is sent).
+        """
+        def count_attachments(message):
+            return (0 if isinstance(message, str)
```
Would do a clear if/elif/else with return statements, currently a bit hard to read.

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+        def count_attachments(message):
+            return (0 if isinstance(message, str)
+                    else sum(count_attachments(part) for part in message.get_payload()) if message.is_multipart()
+                    else 'attachment' in message.get('Content-Disposition', ''))
```
Wouldn't that last one return a boolean? Seems strange.

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+                    else 'attachment' in message.get('Content-Disposition', ''))
+
+        attachment_size_in_bytes = 1024 * 128
+        datas = base64.b64encode((''.join('.' for _ in range(0, attachment_size_in_bytes))).encode())
```
Huh? Are you encoding about 128.000 of "." characters in your test? Seems like a bad idea ^^

Can't we instead either mock "_estimate_mail_size", or better yet mock the "file_size" field?

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+        datas = base64.b64encode((''.join('.' for _ in range(0, attachment_size_in_bytes))).encode())
+        # Define some constant to ease the understanding of the test
+        test_mail_server = self.mail_server_domain_2
+        cst_max_size_always_exceed = 0.1
```
What does "cst" stand for? "constant"? If so, would not specify.

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+        cst_max_size_always_exceed = 0.1
+        cst_max_size_never_exceed = 10
+
+        # cst_max... are used to define max_email_size on the server that is not used with value that would
+        # lead to an incorrect expected_is_links if they were used. We do this to check that the right max
+        # email size is used.
```
🤯

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+        # cst_max... are used to define max_email_size on the server that is not used with value that would
+        # lead to an incorrect expected_is_links if they were used. We do this to check that the right max
+        # email size is used.
+        for (n_attachment, mail_server, default_email_size, max_email_size, business_attachment), expected_is_links in (
```
Note entirely sure why "expected_is_links" is not in the initial tuple, I'm guessing to split data/assertion, but not sure it's worth as it's adding a lot of parenthesis everywhere ^^

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+                ((1, test_mail_server, cst_max_size_never_exceed, 0.1, True), True),
+                # Attachments not linked to a business record are never turned to link
+                ((3, self.env['ir.mail_server'], 0.3, cst_max_size_never_exceed, False), False),
+                ((1, test_mail_server, cst_max_size_never_exceed, 0.1, False), False),
```
Seems like the "constant to use" could be deduced from the couple "business_attachment / expected_is_links".

Would probably simplify a bit the initial dataset and move the complex comment inside the test.

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+                    'attachment_ids': attachments.ids,
+                    'body_html': '<p>Test</p>',
+                    'email_from': 'test@test_2.com',
+                    'email_to': f'u_{mail_idx}_s-{bool(mail_server)}_a-{n_attachment}_l-{expected_is_links}@test.com',
```
Would rather add more parameters to the subTest than having this obscure string that summarizes everything ^^

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+                } for mail_idx in range(2)])
+                mails._send(mail_server=mail_server)
+
+            match_download_links = [
+                re.compile(fr'.*<a.*href.*/web/content/([^?]*)\?.*access_token.*>.*{attachment.name}.*</a>', re.DOTALL)
+                for attachment in attachments]
```
Arf, a regex.

As those are sources of problems of their own, couldn't we just search for the access_token of every attachment and assume that if it's present it means that the link was properly generated?

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+            for mail, outgoing_email in zip(mails, self.emails):
+                message_raw = outgoing_email['message']
+                message_parsed = message_from_string(message_raw)
+                message_cleaned = re.sub(r'[\s=]', '', message_raw)
```
Especially since the first regex is linked to a second one? 🤯

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+                with self.subTest(email_to=mail['email_to']):
+                    if expected_is_links:
+                        self.assertEqual(count_attachments(message_parsed), 0,
+                                         'Attachment have been removed (replaced by download links)')
```
Attachments*

As this message is printed when this does NOT happen, would change "have been" to "should have been removed and replaced by".

To apply to other messages as well.

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+                        self.assertEqual(count_attachments(message_parsed), 0,
+                                         'Attachment have been removed (replaced by download links)')
+                        self.assertTrue(all(pattern.match(message_cleaned) for pattern in match_download_links),
+                                        'All attachment are turned into download links')
```
attachments*

awa-odoo commented on addons/test_mail/tests/test_mail_mail.py:
```diff
@@ -823,6 +830,97 @@ def test_mail_mail_values_unicode(self):
+                        self.assertTrue(all(pattern.match(message_cleaned) for pattern in match_download_links),
+                                        'All attachment are turned into download links')
+                        self.assertTrue(
+                            all(self.env['ir.attachment'].browse(int(pattern.findall(message_cleaned)[0])).access_token
+                                for pattern in match_download_links),
+                            'All download links target attachment with access token')
```
That's a strange one, would say not necessary, I'm not sure how we could randomly add links to attachments that have nothing to do with the email's attachments.

awa-odoo reviewed (commented):
Hello @pyduf 

Some replies to comments as well as remarks on the test.
Goal is as always to simplify and ease reading.

After that we should be close to being able to ship it 🚀 

Thanks & cheers!

pyduf reviewed (commented):
Hello @awa-odoo, 

Thanks for the review and having spotted the error about the remaining attachments. I have updated the code following your comments. 

What’s remaining is answering you about the call to “.read” for  loading the file contents in mail_mail while filtering the record before.

Note that:
–	2 tests have failed on the runbot but as it works locally, I have relaunched the build.
–	For the email size detection, as we move the raise of the exception inside the “try except” that re-raise the exception, we will need an override.

pyduf commented on addons/mail/models/mail_mail.py:
```diff
@@ -442,23 +459,36 @@ def _prepare_outgoing_list(self, recipients_follower_status=None):
+        # load attachment binary data with a separate read(), as prefetching all
+        # `datas` (binary field) could bloat the browse cache, triggering
+        # soft/hard mem limits with temporary data.
+        email_attachments = [(a['name'], a['raw'], a['mimetype'])
```
Hello @awa-odoo , 

I have analysed in details that part and I think you're right. Here is the details:

In any case, we need to loop over the attachments to check if we can/must embed them as query link (to check their global size and check that they are owned by a business record to avoid them to be deleted randomly). So we can’t remove that. Moreover, when embedding we call generate_access_token which also loop over the attachment reading some field.

But following my tests below, the historical “.read” on attachment doesn’t seem to be useful. We could just loop over the record. But following my observation, it doesn’t seems change the number of queries when just looping over the records or reading some specific fields.

Here are my observations:
- When placing a break point in ir_attachment._compute_raw which read the binary data from disk, it is first called when reading explicitly the "raw" with ".read".
- When profiling query for the ".read" statement, no query is generated (when using file system storage, otherwise a query to fetch all the db_datas in one go).
- Replacing the read by a simple loop over the attachments, no additional query is generated as well. And the _compute_raw is also called from there and not before.
- On the contrary, attachments are read when filtering them above as you mention but the content of the file are not loaded (as it is a compute that need an access to it to be triggered and db_datas is a binary field with preftech=False).
- To generate the embedded links, we need to call generate_access_token which loop over attachment while reading some field (similarly to what filter does but potentially on a subset of records).

As far as I understand, binary data can be stored in multiple location, the default one being the file system. But there is also a column "db_datas" that can hold binary data.
If I configure odoo to use it. It doesn't change the behavior: when not reading explicitly the content of the files, it is not read in the cache. This is because binary field have the "prefetch" attribute to False by default.
So it seems that unless the field "raw" (or "db_datas") is accessed explicitly, the file content won't be read. So it seems that using read instead of looping on the attachment records is not useful (and could maybe in some case generate an additional query). Moreover, when debugging the ".read", the db_datas are inserted in the cache.

In conclusion, using read instead of looping over the records seems unnecessary but I might missing something. But keeping that historical statement doesn’t seems to hurt either. 

Should I replace the read by a simple loop on records ? What do you think ?

awa-odoo replied:
Wow, thank you for the detailed analysis 😄 

I have traced back the initial issue to (10 years ago 👴):
https://github.com/odoo/odoo/commit/61d212298773d0c49048fde9f0a1af9f1a2d1df1

Looks like it's more a problem of "don't try to load the attachments of all emails being sent at once". .
So I would say let's not mess with it and keep it as is.

awa-odoo commented on odoo/addons/base/models/ir_mail_server.py:
```diff
@@ -248,6 +263,12 @@ def test_smtp_connection(self):
+                if autodetect_max_email_size:
+                    max_size = smtp.esmtp_features.get('size')
+                    if not max_size:
+                        raise UserError(_('The server "%(server_name)s" doesn\'t return the maximum email size.',
```
Arf, looks like check style is still unhappy about this one.

How about we remove this part:
```py
except UserError:
    raise
```

And modify this part:
```py
except Exception as e:
    if not isinstance(e, UserError):
        _logger.warning("Connection test on %s failed with a generic error.", server, exc_info=True)
        raise UserError(_("Connection Test Failed! Here is what we got instead:\n %s", e)) from e
```

Would that do the trick and achieve the same behavior?

awa-odoo commented on addons/mail/models/mail_mail.py:
```diff
@@ -442,23 +459,36 @@ def _prepare_outgoing_list(self, recipients_follower_status=None):
+                attachments_links = self.env['ir.qweb']._render('mail.mail_attachment_links',
+                                                                {'attachments': record_owned_attachments})
+                body = tools.append_content_to_html(body, attachments_links, plaintext=False)
+                attachments = attachments - record_owned_attachments
```
Could use `-=` syntax.

awa-odoo reviewed (commented):
Hello @pyduf 

Some replies as well as a suggestion to try to make ci/style happy and then we should be good to go as far as I'm concerned.

Thanks again for your work & cheers!

pyduf reviewed (commented):
Hello @awa-odoo , 

Thanks again for the reviews. 

I have updated the code with the last changes and the runbot is green except for the style for which we need an exception (I don't remember how to ask, sorry).

awa-odoo commented:
@robodoo override ci/style

Reason: New code matches usage of all previous exceptions thrown, will have to rework the whole method in a future task to improve that part.

(Hopefully this is the correct syntax).

awa-odoo commented:
@robodoo override=ci/style

Reason: New code matches usage of all previous exceptions thrown, will have to rework the whole method in a future task to improve that part.

(Hopefully this is the correct syntax).

robodoo commented:
I'm sorry, @awa-odoo: you are not allowed to override this status.

tde-banana-odoo commented:
@robodoo override=ci/style

Reason: New code matches usage of all previous exceptions thrown, will have to rework the whole method in a future task to improve that part.

awa-odoo commented:
@robodoo r+

Julien00859 commented:
Sorry to come after the battle... but could we make it explicit in the source code that the max size is in MiB? There are several other ICP limits inside core/base and they are in bytes, not MiB, I'm a bit sad that we broke the consistency here :(. An help attribute on the `ir.mail_server.max_email_size` would be welcome too :pray: 

Also to :nerd_face: a bit 1 MB = 1000 kB = 1.000.000 B, 1 MiB = 1024 kiB = 1.048.576 B. Don't let `odoo.tools.human_size` fool you (frankly that function should be deprecated).
